tracks/javascript/exercises/list-ops/mentoring.md

```diff
@@ -0,0 +1,304 @@
+
+Most students will actually still use a backing array for the values but have
+various ways of not using the built in functions to process them. There are (at
+least) two correct ways of implementing this using only array destructuring to
+go from the test array to the internal representation and from the internal
+representation back to `values`.
+
+> *Note*: You don't need **any** function from the `Array.prototype`, and only
+> need the language feature to build construct and desconstruct arrays via
+> destructuring as well as the array literal `[]`.
+>
+> *Note*: most implementations using `for` loops, also use `.length` on the
+> backing array, or the `array[indexer]`. You don't need either!
+
+#### Functional data type without backing array
+
+This is a representation of one of the most fundamental *cons* tructs in most
+functional languages called [`Cons`](https://en.wikipedia.org/wiki/Cons). It
+is actually a way to build a linked-list.
+
+```javascript
+Cons.fromArray([1, 42, 'foo'])
+// => Cons(1, Cons(42, Cons('foo', Null)))
+
+Cons.fromArray()
+// => Null
+
+Cons(1, Cons(42, Cons('foo', Null))).values
+// => [1, 42, 'foo']
+
+Null.values
+// => []
+```
+
+The `Null` object is a convenience object to replace `null` or `undefined` at
+the end of the list and allows for code without a any branches (exception is
+the conversion code).
+
+```javascript
+const Null = {
+  get value() { return undefined },
+  get next() { return this },
+  get values() { return [] },
+
+  get() { return this.value },
+  push(item) { return new Cons(item, this) },
+  length() { return 0 },
+  append(other) { return other },
+  concat() { return this },
+  forEach() { /* done */ },
+  foldl(_, initial) { return initial },
+  foldr(_, initial) { return initial },
+  filter() { return Null },
+  reverse() { return this },
+  map() { return this },
+}
+
+class Cons {
+  static fromArray([head, ...tail]) {
+    if (head === undefined) {
+      return Null
+    }
+
+    return new Cons(head, Cons.fromArray(tail || []))
+  }
+
+  constructor(value, next = Null) {
+    this.value = value
+    this.next = next
+  }
+
+  get values() {
+    return [this.value, ...this.next.values]
+  }
+
+  get(i) {
+    return i === 0
+      ? this.value
+      : this.next.get(i - 1)
+  }
+
+  push(item) {
+    this.next = this.next.push(item)
+    return this
+  }
+
+  length() {
+    return 1 + this.next.length()
+  }
+
+  append(other) {
+    return other.foldl((result, item) => result.push(item), this)
+  }
+
+  concat(others) {
+    return others.foldl((result, other) => result.append(other), this)
+  }
+
+  foldl(callback, initial = undefined) {
+    return this.next.foldl(callback, callback(initial, this.value))
+  }
+
+  forEach(callback) {
+    this.foldl((_, item) => callback(item))
+  }
+
+  foldr(callback, initial = undefined) {
+    return callback(this.next.foldr(callback, initial), this.value)
+  }
```

Comment:
Yay! `foldl` and `foldr` are super elegant now 👍 I guess we don't need `forEach` then?

Reply:
We don't, but I keep it because it's valid. Using JUST foldl is fine, but using for each is more declarative I think :)

Reply:
Yes, but the problem statement doesn't require `forEach`. So should we keep it in expected solution?

Reply:
I don't like mentors suggesting to remove `forEach`, because using `forEach` to `fold` is only minorly less nice than `folding` to `forEach` (if you get what I mean). Yes, the current solution is optimal, but having a `forEach` is okay I guess 👍 

How do you feel about this?

Reply:
Right. But I think we should not add `forEach` in the expected solution. We can mention in the notes that someone can optionally choose to use `forEach` instead of fold and that is totally fine.

Reply:
Got it. I'll change it!